src/scumm-8.p8

```diff
@@ -2605,84 +2317,146 @@ function recalc_bounds(obj, w, h, cam_off_x, cam_off_y)
 end
 
 
--- 
--- a* pathfinding functions 
+--
+-- a* pathfinding functions
 --
 
 function find_path(start, goal)
- local frontier, came_from, cost_so_far, lowest_dist, lowest_dist_node  = {}, {}, {}, nil, nil
+	local frontier, came_from, cost_so_far, lowest_dist, lowest_dist_node, current = {}, {}, {}
  insert(frontier, start, 0)
- came_from[vectoindex(start)] = nil
+	-- came_from[vectoindex(start)] = nil
  cost_so_far[vectoindex(start)] = 0
 
  while #frontier > 0 and #frontier < 1000 do
  	-- pop the last element off a table
-	local top = frontier[#frontier]
+		current = frontier[#frontier][1]
 	del(frontier,frontier[#frontier])
-	current = top[1]
 
-  if vectoindex(current) == vectoindex(goal) then
-   break
-  end
+		if (vectoindex(current) == vectoindex(goal)) break
 
   --local neighbours = getneighbours(current)
 	local neighbours = {}
 	for x = -1, 1 do
-		for y = -1, 1 do
-			if x == 0 and y == 0 then 
-				--continue 
-			else
-				local chk_x = current[1] + x
-				local chk_y = current[2] + y
+			for y = -1, 1, x == 0 and 2 or 1 do
+				local chk_x, chk_y = current[1] + x, current[2] + y
 
-				-- diagonals cost more
-				if abs(x) != abs(y) then cost=1 else cost=1.4 end
-				
 				if chk_x >= room_curr.map[1] and chk_x <= room_curr.map[1] + room_curr.map_w
 				 and chk_y >= room_curr.map[2] and chk_y <= room_curr.map[2] + room_curr.map_h
 				 and is_cell_walkable(chk_x, chk_y)
 				-- squeeze check for corners
-				 and ((abs(x) != abs(y)) 
-						or is_cell_walkable(chk_x, current[2]) 
+				 and ((abs(x) != abs(y))
+						or is_cell_walkable(chk_x, current[2])
 						or is_cell_walkable(chk_x - x, chk_y)
-						or enable_diag_squeeze) 
+						or enable_diag_squeeze)
 				then
-					-- add as valid neighbour
-					add( neighbours, {chk_x, chk_y, cost} )
+					-- process a valid neighbor
+					local next = {chk_x, chk_y}
+					local nextindex = vectoindex(next)
+					local new_cost = cost_so_far[vectoindex(current)] + (x * y == 0 and 1 or 1.414) -- diagonals cost more
+
+					if not cost_so_far[nextindex] or new_cost < cost_so_far[nextindex] then
+						cost_so_far[nextindex] = new_cost
+						local h = max(abs(goal[1] - chk_x), abs(goal[2] - chk_y)) + min(abs(goal[1] - chk_x), abs(goal[2] - chk_y)) * .414
+						insert(frontier, next, new_cost + h)
+						came_from[nextindex] = current
+
+						if not lowest_dist or h < lowest_dist then
+							lowest_dist = h
+							lowest_dist_node = nextindex
+							lowest_dist_neigh = next
+						end
+					end
```

Comment:
I couldn't see or think of any reason to collect all of the neighbors rather than just processing them as they are enumerated. (and of course I forgot to remove the `neighbors` declaration)